### PR TITLE
support ref intent on gpuized forall w/ AMD

### PR DESCRIPTION
**Pull request by stonea**
This PR extends the work from #25594 so that `ref` intent scalars will work with `forall` loops on AMD GPUs (the prior PR got it to work for NVIDIA).

Since AMD GPUs (or at least the ones we're using today) don't have universal addressing we need to convert the host pointer to the `ref` intent'd parameter into a "device pointer" before it can be used by the device. This PR performs this step.

Notes to reviewers:

For reasons I don't fully understand, in order to get this to work the thing I pass to the `cfg_add_direct_param` function (to add the ref intent'd param as a kernel argument) needs to be a double pointer (with the pointee being the aforementioned device pointer). To get this all to work I create an array of void pointers and then a second array where each element is a pointer to each element in the first. This lets us avoid malloc'ing each double pointer individually but, as can be seen, the code has a bit of pointer juggling. I'm not totally convinced my approach here is the most optimal thing to do but it seems to work.

### CODE REVIEW

File: runtime/src/chpl-gpu.c [hunk 1]
```diff
@@ -257,7 +257,8 @@ typedef struct kernel_cfg_s {
   void*** kernel_params;
   int n_host_registered;
   int cur_host_registered_var;
-  void** host_registered_vars;
+  void** host_registered_var_boxes;
+  void*** host_registered_vars;
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:261]: I thrive around `void***`s...

File: runtime/src/gpu/amd/gpu-amd.c [hunk 2]
```diff
@@ -391,15 +391,15 @@ bool chpl_gpu_impl_can_sort(void){
...
+void* chpl_gpu_impl_host_register(void* var, size_t size) {
+  hipHostRegister(var, size, hipHostRegisterPortable);
+  void *dev_var;
+  hipHostGetDevicePointer(&dev_var, var, 0);
+  return dev_var;
 }
 
 void chpl_gpu_impl_host_unregister(void* var) {
-  // hipHostUnregister(var);
+  hipHostUnregister(var);
```
Comment (e-kayrakli) [inline runtime/src/gpu/amd/gpu-amd.c:402]: These should have `ROCM_CALL` around them. I also see that they are missing in CUDA versions. Could you add them while here?

File: runtime/src/chpl-gpu.c [hunk 3]
```diff
@@ -700,9 +708,12 @@ void chpl_gpu_arg_reduce(void* _cfg, void* arg, size_t elem_size,
 
 void chpl_gpu_arg_host_register(void* _cfg, void* arg, size_t size) {
   kernel_cfg* cfg = (kernel_cfg*)_cfg;
-  cfg_add_direct_param(cfg, arg);
-  chpl_gpu_impl_host_register(*((void**)arg), sizeof(char)*8);
-  CHPL_GPU_DEBUG("\tAdded ref intent param (at %d): %p\n", cfg->cur_param,  arg);
+
+  void *dev_arg = chpl_gpu_impl_host_register(*((void**)arg), size);
+  *(cfg->host_registered_vars[cfg->cur_host_registered_var]) = dev_arg;
+  cfg_add_direct_param(cfg, cfg->host_registered_vars[cfg->cur_host_registered_var]);
```
Comment (e-kayrakli) [inline runtime/src/chpl-gpu.c:714]: The parameters we pass to the kernel launchers are addresses of things that we want to pass as arguments. `dev_arg` is the thing we want to pass, so we actually have to pass its address. Here, you made `cfg->host_registered_vars` an array that can store such addresses. You couldn't have passed `&dev_arg` as it'll go out of scope.

All this is to say that this makes sense but annoyingly convoluted. We have a similar pattern elsewhere in the runtime I believe. Maybe the `offload` version of arg passing?

File: runtime/src/gpu/nvidia/gpu-nvidia.c [hunk 4]
```diff
@@ -379,8 +379,9 @@ bool chpl_gpu_impl_can_sort(void){
   return true;
 }
 
-void chpl_gpu_impl_host_register(void* var, size_t size) {
+void* chpl_gpu_impl_host_register(void* var, size_t size) {
   cuMemHostRegister(var, size, CU_MEMHOSTREGISTER_PORTABLE);
+  return var;
```
Comment (e-kayrakli) [inline runtime/src/gpu/nvidia/gpu-nvidia.c:384]: Please add `CUDA_CALL`s to this function and the symmetric unregister one.

**Review by e-kayrakli — APPROVED** [review]

**Comment by bradcray** [comment]
Great!

**Pull request merged**